### PR TITLE
Track pending JS calls to allow for monitoring of bridge traffic

### PR DESCRIPTION
**Pull request by matthargett**
Pulled from RN Android: 
https://github.com/facebook/react-native/blob/4466b6fa7ce5325cac3233a23b10ffcb4ff038a7/ReactAndroid/src/main/java/com/facebook/react/bridge/CatalystInstanceImpl.java#L469

It enables some shutdown safety and performance monitoring hooks like this one: 
https://github.com/facebook/react-native/blob/26684cf3adf4094eb6c405d345a75bf8c7c0bf88/ReactAndroid/src/main/java/com/facebook/react/modules/debug/FpsDebugFrameCallback.java

(I'd prefer to do an impl of FpsDebugFrameCallback in a separate PR.)

Thanks to my colleagues @takesui and @yuichiando for doing the port, integration, and testing work!

### CODE REVIEW

**Comment by matthargett** [comment]
Closes #1694

**Comment by rozele** [comment]
This is related to https://github.com/Microsoft/react-native-windows/issues/31

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs [hunk 1]
```diff
@@ -106,13 +110,32 @@ public void Dispose()
             _jsExecutor.Dispose();
         }
 
-        private void ProcessResponse(JToken response)
+        private void ProcessResponse(JToken response, bool isEndOfBatch = false)
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: What's the point of this argument?

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs [hunk 2]
```diff
@@ -147,7 +176,18 @@ private void ProcessResponse(JToken response)
...
 
-                _reactCallback.OnBatchComplete();
+                if (isEndOfBatch)
+                {
+                    // onBatchComplete will be called on the native (module) ActionQueue, but
+                    // decrementPendingJSCalls will be called sync. Be aware that the bridge may still
+                    // be processing native calls when the bridge idle signaler fires.
+                    if (localBatchHadNativeModuleCalls)
+                    {
+                        _reactCallback.OnBatchComplete();
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: OnBatchComplete is a critical piece of infrastructure (it's how the UIManagerModule batches view operations), I'm concerned with this now being hidden behind conditionals.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 3]
```diff
@@ -14,10 +14,13 @@
 using System.IO;
 using System.Threading;
 using System.Threading.Tasks;
+using ReactNative.Chakra;
 using static System.FormattableString;
 
 namespace ReactNative.Bridge
 {
+    public delegate void OnTransitionToBridgeHandler(bool busy);
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: In general, we try to avoid multiple classes (even delegates) in one file.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 4]
```diff
@@ -33,6 +36,13 @@ class ReactInstance : IReactInstance
 
         private bool _initialized;
 
+        public event OnTransitionToBridgeHandler OnTransitionToBridgeHandler;
+
+        private bool BridgeBusy
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: Why one event instead of two (one for busy, one for idle)?

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs (hunk 1 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: Seems like it might be in preparation for the `nativeFlushQueueImmediate` behavior. If that's the case, please defer these changes until that feature is pulled in as well. Ideally, that feature should also be done in a separate PR.

Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: I did try to do them separately before, but the changes were fairly intertwined. I can bring in nativeFlushQueueImmediate as well and rename this PR, if you like. lmk.

Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: Seems like that `isEndOfBatch` parameter is the only relic of the `nativeFlushQueueImmediate` feature (that I can see).

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 4 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: great question, I'm not sure why we deviated from Java there. I'll change it to be congruent with Java for easier back/forward porting of fixes/enhancements with upstream.

Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: #fixed

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 3 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: #fixed

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs (hunk 2 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: I tweaked the comments to best reflect the combination of our (existing) structural differences, and moved the setting of the state to be closer to RN Android/cxxreact. I tried inverting the guards and other things, but the code didn't end up any cleaner and had the same number of conditionals. I'm open to specific suggestions.

**Comment by rozele** [comment]
This needs rebasing.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 5]
```diff
@@ -33,6 +33,9 @@ class ReactInstance : IReactInstance
 
         private bool _initialized;
 
+        public event TransitionToBridgeIdleCallback OnTransitionToBridgeIdleCallback;
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: >TransitionToBridgeIdleCallback [](start = 21, length = 30)

Any reason we can't just use `Action` here? Or, switch to the more common usage of `EventHandler<SomeEventArgs>`?

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 6]
```diff
@@ -175,6 +178,8 @@ public void InvokeCallback(int callbackId, JArray arguments)
             IsDisposed = true;
 
             await QueueConfiguration.NativeModulesQueue.RunAsync(_registry.NotifyReactInstanceDispose).ConfigureAwait(false);
+            // RN Android is structured differently and notifies idle listeners here, should we do the same?
+            // CatalystInstanceImpl.java#L331
             await QueueConfiguration.JavaScriptQueue.RunAsync(() => _bridge?.Dispose()).ConfigureAwait(false);
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: A long time ago, there was a TODO for this :-):
[https://github.com/Microsoft/react-native-windows/blame/1ef3fb438907545d2008ce51085afc5737e75f28/ReactNative/Bridge/CatalystInstance.cs#L102](https://github.com/Microsoft/react-native-windows/blame/1ef3fb438907545d2008ce51085afc5737e75f28/ReactNative/Bridge/CatalystInstance.cs#L102)

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs [hunk 7]
```diff
@@ -311,6 +317,24 @@ public void OnBatchComplete()
                     }
                 }
             }
+
+            public void IncrementPendingJSCalls()
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: >IncrementPendingJSCalls [](start = 24, length = 23)

In almost all cases, we haven't used acronyms for JS (I think originally due to some StyleCop rule we had been using that didn't allow acronyms, we no longer do this). But we should be consistent nonetheless. Since these are called `BridgeIdle` and `BridgeBusy`, I'd be fine with calling this `IncrementPendingBridgeCalls`, it's internal code anyway. No one is ever really expected to implement `IReactCallback`.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 5 above)
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: >OnTransitionToBridgeIdleCallback [](start = 52, length = 32)

.NET events typically don't have an `On` prefix, and likely not a `Callback` suffix, since both of these are implied by the nature of the event. Can we just call these `BridgeIdle` and `BridgeBusy`?

Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: >public event [](start = 8, length = 12)

Do / should these events go in the `IReactInstance` interface? I don't know the answer to this.

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs [hunk 8]
```diff
@@ -56,6 +56,7 @@ public class ReactBridge : IReactBridge
         public void CallFunction(string moduleName, string method, JArray arguments)
         {
             var response = _jsExecutor.CallFunctionReturnFlushedQueue(moduleName, method, arguments);
+            _reactCallback.IncrementPendingJSCalls();
             ProcessResponse(response, true);
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: Why are we incrementing after the _synchronous_ call to JavaScript is completed? Shouldn't this occur before the JavaScript call?

This may just be a problem with the name of the function (see other comment).  When I read this code, it makes me think I should increment the pending JS calls _before_ doing any work in JS.

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs [hunk 9]
```diff
@@ -197,6 +198,301 @@ public void ReactBridge_ValidJavaScriptResponse()
             }
         }
 
+        [Test]
+        public async Task SingleReturnCallNoArgs()
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: >SingleReturnCallNoArgs [](start = 26, length = 22)

Looks like there is a naming convention for tests that prefixes the class name being tested.

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs [hunk 10]
```diff
@@ -197,6 +198,301 @@ public void ReactBridge_ValidJavaScriptResponse()
...
+
+                bridge.CallFunction("module", "method", new JArray());
+
+                await Task.Run(new Action(countdownEvent.Wait));
+                Assert.That(called, Is.EqualTo(1));
+            }
+        }
+
+        [Test]
+        public async Task ReturnMultipleParams()
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: >ReturnMultipleParams [](start = 26, length = 20)

It looks like these tests may belong in a separate PR. I don't see anything specific to bridge idle/busy here.

Comment (rozele) [reply ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: That being said, we should add some tests :-)

---
In reply to: [185373727](https://github.com/Microsoft/react-native-windows/pull/1726#discussion_r185373727) [](ancestors = 185373727)

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 5 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: Since it's just a binary, what would the additional data be to justify an Action or an EventHandler + new Argument type?

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 6 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: better late than never ;) do you agree that the structure is different, so where we notify is appropriate given the divergence?

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 7 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: I was just following upstream interface names to make tracking future changes easier, happy to change it.

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs (hunk 8 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: I went back and forth with this before committing, using upstream as a reference. I'm okay to change it.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 5 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: I was trying to track RN Android, and reconcile with a multi-Instance app architecture.

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs (hunk 9 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: that's a deficiency of the Test Explorer UI that's being resolved in an upcoming VS2017 update (or maybe it's already there and not on by default?). Every other test runner (include dotnet CLI) includes the full type of the enclosing test fixture.

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs (hunk 10 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: Wanted to pin the behavior first, will add appropriate assertions to the new tests in the next step.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 5 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: @matthargett - I'm fine with just using a non-generic `Action` then so we don't need to invent a new data type, or a non-generic `EventHandler` delegate type.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 6 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: I think this is just a convention that one final idle callback should occur when the `ReactInstance` is being torn down.

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 7 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: There are a few cases where we diverge from upstream to make things feel more "C#-y"

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs (hunk 8 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: #fixed

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 7 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: #fixed

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 6 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: #fixed

File: ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs (hunk 5 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: #fixed

Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: changed to an Action as per above and named according to other similar Actions in codebase.

Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactInstance.cs]: Actually, looking at RN Android and another implementation -- yes. Otherwise, you'd end up casting the IReactInstance getter on the Context to the concrete. I'll fix this next.

**Comment by rozele** [comment]
>         public async Task ReactInstance_GetModules()

Instead of piggy backing on existing tests, would it make sense to make standalone tests for the idle/busy behavior?

---
Refers to: ReactWindows/ReactNative.Tests/Bridge/ReactInstanceTests.cs:17 in 196e157. [](commit_id = 196e157e14bb710a6630b4e9c89929f81b40f31e, deletion_comment = False)

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs (hunk 9 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: Still nice to remain consistent for now...

---
In reply to: [185671193](https://github.com/Microsoft/react-native-windows/pull/1726#discussion_r185671193) [](ancestors = 185671193)

File: ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs [hunk 11]
```diff
@@ -17,22 +18,52 @@ namespace ReactNative.Tests.Bridge
...
     public class ReactBridgeTests
     {
+        private JToken _response;
+        private List<(int moduleId, int methodId, JArray args)> _receivedCallbacks;
+        private AutoResetEvent _eventHandler;
+        private MockReactCallback _mockReactCallback;
+        private MockJavaScriptExecutor _mockJavaScriptExecutor;
+        private IActionQueue _nativeModulesQueueThread;
+        private ReactBridge _reactBridge;
+
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared.Tests/Bridge/ReactBridgeTests.cs]: Can we make these changes to tests in a different PR please?

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs [hunk 12]
```diff
@@ -115,8 +118,17 @@ private void ProcessResponseImmediate(JToken response)
...
-            // Gets the number of native module calls and validates the batch.
-            var moduleCallCount = GetModuleCallCount(response);
+            int moduleCallCount;
+
+            try
+            {
+                moduleCallCount = GetModuleCallCount(response);
+            }
+            catch (Exception)
+            {
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: I don't think we need to catch an exception here. It's not done in the ReactCommon bridge: [https://github.com/facebook/react-native/blob/master/ReactCommon/cxxreact/MethodCall.cpp#L23-L31](https://github.com/facebook/react-native/blob/master/ReactCommon/cxxreact/MethodCall.cpp#L23-L31)

File: ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs [hunk 13]
```diff
@@ -60,5 +61,15 @@ public interface IReactInstance : IAsyncCancelable
         /// <typeparam name="T">Type of native module.</typeparam>
         /// <returns>The native module instance.</returns>
         T GetNativeModule<T>() where T : INativeModule;
+
+        /// <summary>
+        /// Fires when there were, but are now no, pending bridge calls
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs]: >Fires when there were, but are now no [](start = 12, length = 37)

"Fires when the number of pending bridge calls goes to zero."

File: ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs [hunk 14]
```diff
@@ -60,5 +61,15 @@ public interface IReactInstance : IAsyncCancelable
...
+        /// </summary>
+        event Action BridgeIdle;
+
+        /// <summary>
+        /// Fires when there are none, but now are, pending bridge calls
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs]: >Fires when there are none, but now are, pending bridge calls [](start = 12, length = 60)

"Fires when the bridge is no longer idle."

File: ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs (hunk 13 above)
Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/IReactInstance.cs]: Just suggestions, the comments are a bit confusing.

---
In reply to: [186073353](https://github.com/Microsoft/react-native-windows/pull/1726#discussion_r186073353) [](ancestors = 186073353)

File: ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs (hunk 12 above)
Comment (matthargett) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: I was actually going to follow up there with a PR as well, but I can do it upstream and then bring it back here if that's preferrable.

Comment (rozele) [reply ReactWindows/ReactNative.Shared/Bridge/ReactBridge.cs]: If it's truly a bug... there are pretty strong guarantees about the contract of the `BatchedBridge`, I just assumed that the exceptions thrown might as well have been a Debug assertion.

**Comment by matthargett** [comment]
Removed offending tests and left them to the fork. If there's no additional feedback, please merge. Thanks!
